lib/puppet/util/tagging.rb

```diff
@@ -40,7 +40,7 @@ def tag(*ary)
   # since that results in testing the same string twice
   #
   def tag_if_valid(name)
-    if name.is_a?(String) && name =~ ValidTagRegex
+    if name.is_a?(String) && name =~ ValidTagRegex && Puppet::Resource::MAIN != name
```

Comment:
As @hlindberg said, "we don't know all the places that may matter." That is exactly my concern. I don't want to play a game of whack-a-mole, trying to figure out all of the places we need to check if it's the `main` class or not.

However, if we can prevent a user from ever creating a main class (like you've done in commit 9771456, then do we need to add any checks at all? For example, can we drop the `&& Puppet::Resource::MAIN != name` conditional above, since a user can't add a `main` class, so we don't have to worry about it being tagged?